### DISCUSSION TITLE
Integration with editors / IDEs

### DISCUSSION THREAD

**Discussion by eproxus** [jj-vcs/jj#4424] [Ideas]
I tried to find info on this but couldn't find anything (so apologies if it has been brought up before): is there any way to integrate with editors, tools or IDEs that are already Git aware? 

I'm mostly interested in my editor being able to show the current workspace diff and being able to look at the log and blame. Since `jj` uses Git under the hood, I was hoping that there would be some Git repository hidden somewhere which could be used. I found `.jj/store/repo/git` but that seems to stay on `main` regardless of which branch the `jj` checkout is on.

It would be cool if there would be a way for `jj` to export some `.git` directory that somehow represents the current working state of the `jj` repo, but I realize this might be extremely hard or impossible (e.g. see the work that has gone into GitButler to represent multiple branches active in the same repo at the same time).

Does this idea even remotely make sense, or is it just a pipe dream and we should wait for native `jj` integration instead?

**Comment by martinvonz** [comment]
This is one common reason people use [colocated repo](https://martinvonz.github.io/jj/latest/git-compatibility/#co-located-jujutsugit-repos) (which should perhaps be called "colocated workspaces").

**Reply by eproxus** [reply]
That is great to hear, will try it out! A bit of feedback: I search a long time for "editor integration" and "IDE support" and various other phrases in that vein, but didn't find anything. Perhaps some more keywords like that could be added so it would be easier to find?

**Reply by cormacrelf** [reply]
It's missing one thing if you want to call it colocated workspaces -- `jj workspace add ../second` should make the second workspace also appear as a git repo. Currently you can't do this if you use Nix flakes / direnv, because it will no longer respect your .gitignore.

**Reply by ilyagr** [reply]
This might be worth creating an FR about. It might not be trivial to implement, but could be doable. I expect some diving into libgit2's and/or gitoxide's support for worktrees would be required, and we'd have to hope their support is good enough.

The most natural way this could go is if we could make the "co-located workspace" also a Git worktree. This would mean that `jj git export` and `jj git import` would have to learn to understand worktrees.

**Reply by cormacrelf** [reply]
https://github.com/martinvonz/jj/issues/4436
